File: LL/basic.py
class ListNode:
	def __init__(self, val): # Constructor to initialize the node object
		self.val = val
		self.next = None

class LinkedList:
	def __init__(self): # Constructor to initialize head
		self.head = None

	def insertAtStart(self, node):
		new_node = ListNode(node) 
		new_node.next = self.head # Make next of new node as head 
		self.head = new_node # move the head to point to the new node

	def insertAfter(self, prev, node):
		if prev == None:
			return
		new_node = ListNode(node)
		new_node.next = prev.next
		prev.next = new_node		

	def deleteNode(self, node):
		temp = self.head # store head node

		# If head node itself holds the key to be deleted 
		if temp is not None and temp.val == node:
				self.head = temp.next # change head
				temp = None	# free old head
				return 

		# Search for the key to be deleted, keep track of the 
		# previous node as we need to change 'prev.next' 
		while temp is not None and temp.data != node:
			prev = temp
			temp = temp.next

		# if key was not present in linked list
		if temp == None:
			return

		#unlink the node from LL
		prev.next = temp.next
		temp = None				

	def deleteNodeAtPos(self, pos):
		if self.head == None: # if LL is empty
			return

		temp = self.head # Store head node

		# if head need to be removed
		if pos == 0:	
			self.head = temp.next
			temp = None
			return

		# Find previous node of the node to be deleted
		for i in range(pos-1 ): 
			temp = temp.next
			if temp is None: 
				break	

		# if pos is more than no of nodes
		if temp is None or temp.next is None:
			return	

		next = temp.next.next
		temp.next = None # free memory
		temp.next = next	

	def append(self, node):
		new_node = ListNode(node)
		last = self.head

		# If the Linked List is empty, then make the new node as head 
		if self.head == None:
			self.head = new_node
			return

		while(last.next != None): # traverse till the last node
			last = last.next

		last.next = new_node # Change the next of last node
		return	


	def search(self, li, key):
		if li == None:
			return False
		if li.val == key:
			return True
		return self.search(li.next, key)

	def printList(self):
		temp = self.head
		while(temp):
			print("%d" %(temp.val), end = " ")	
			temp = temp.next
		print(" ")	

if __name__ == '__main__':
	llist = LinkedList()
	llist.insertAtStart(6)
	llist.insertAtStart(4)
	llist.insertAtStart(3)
	llist.insertAtStart(2)
	llist.insertAtStart(1)
	llist.printList()
	llist.append(7)
	llist.printList()
	llist.insertAtStart(0)
	llist.printList()
	llist.insertAfter(llist.head.next.next.next.next, 5) # insert after 4
	llist.deleteNode(0)
	llist.printList()
	llist.deleteNodeAtPos(5)
	llist.printList()
	if llist.search(llist.head, 3) == True:
		print("yes")
	else:
		print("no")	

File: LL/basic.cpp
#include<iostream>
#include<vector>
#include<cmath> 
#include<algorithm>
#include<set>
#include<string>
#include<map>
#include<unordered_map>
#include<cstring>
#include<stack>
#include<queue>

using namespace std;

#define ll long long

const int INF = 1e9+7;
const int mod = 1e9+7;
const int MAX = 1e5+5;

struct ListNode{
    int val;
    ListNode* next;
    ListNode(int x): val(x), next(NULL){}
};

void append(ListNode** head, int num){
    ListNode* new_node = new ListNode(num);
    ListNode* last = *head;

    // If the Linked List is empty, then make the new node as head 
    if(*head == NULL){
        *head = new_node;
        return;
    }   

    while(last->next != NULL) // traverse till the last node
        last = last->next;

    last->next = new_node; // Change the next of last node
    return;
}

void insertAtStart(ListNode** head, int num){
    ListNode* new_node = new ListNode(num);

    new_node->next = *head;   // Make next of new node as head 

    *head = new_node; // move the head to point to the new node
}

void insertAfter(ListNode* prev, int num){
    if(prev == NULL)
        return;

    ListNode* new_node = new ListNode(num);
    // new_node->val = num;
    new_node->next = prev->next; // Make next of new node as next of prev_node
    prev->next = new_node; // move the next of prev_node as new_node
}


void deleteNode(ListNode** head, int key){
    ListNode* temp = *head, *prev;

    // If head node itself holds the key to be deleted
    if(temp != NULL && temp->val == key){
        *head = temp->next; // changed head
        free(temp); // free old head
        return;
    }

    // Search for the key to be deleted, keep track of the
    // previous node as we need to change 'prev->next' 
    while(temp != NULL && temp->val != key){
        prev = temp;
        temp = temp->next;
    }

    // if key is not present in LL
    if(temp == NULL)
        return;

    prev->next = temp->next; // unlink the node from LL
    free(temp); // free memory
}

void deletedNodeAtPos(ListNode **head, int pos){
    // if LL is empty
    if(*head == NULL)
        return;

    ListNode *temp = *head;

    if(pos == 0){ // if head to be removed
        *head = temp->next; // change head
        free(temp);
    }

    // find prev node of the node to be deleted
    for(int i = 0; temp != NULL && i < pos-1; ++i){
        temp = temp->next;
    }

    // if pos is more than the no of nodes
    if(temp == NULL || temp->next == NULL)
        return;

    ListNode* next = temp->next->next;
    free(temp->next);
    temp->next = next;  
}

bool search(ListNode* head, int key){
    if(head == NULL)
        return false;
    if(head->val == key)
        return true;
    return search(head->next, key);
}

void printList(ListNode* node){
    while(node != NULL){
        cout << " " << node->val;
        node = node->next;
    }
    cout << endl;
}

int main(){
	//freopen("tree.in","r",stdin);
    //freopen("tree.out","w",stdout);
    //freopen("input.txt", "r", stdin);
    //freopen("out.txt", "w", stdout);
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    ListNode *head = NULL;
    insertAtStart(&head, 6);
    insertAtStart(&head, 4);
    insertAtStart(&head, 3);
    insertAtStart(&head, 2);
    insertAtStart(&head, 1);
    printList(head);
    append(&head, 7);
    printlist(head);
    insertAfter(head->next->next->next, 5);    
    printList(head);
    insertAtStart(&head, 0);
    printList(head);
    deleteNode(&head, 0);
    printList(head);
    deletedNodeAtPos(&head, 5); // pos
    printList(head);
    cout << (search(head, 3)?"yes " : " no") << endl;

}